include/aws/common/string.h

```diff
@@ -0,0 +1,100 @@
+#ifndef AWS_COMMON_STRING_H
+#define AWS_COMMON_STRING_H
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/common.h>
+#include <aws/common/byte_buf.h>
+#include <string.h>
+
+/**
+ * Represents an immutable string holding either text or binary data. If the string is in constant
+ * memory or memory that should otherwise not be freed by this struct, set allocator to NULL
+ * and destroy function will be a no-op.
+ *
+ * This is for use cases where the entire struct and the data bytes themselves need to be
+ * held in dynamic memory, such as when held by a struct aws_hash_table. The data bytes
+ * themselves are always held in contiguous memory immediately after the end of the
+ * struct aws_string, and the memory for both the header and the data bytes
+ * is allocated together.
+ *
+ * Use the aws_string_bytes function to access the data bytes.
+ */
+struct aws_string {
+    struct aws_allocator * allocator;
+    size_t len;
+};
+
+static inline const uint8_t * aws_string_bytes(const struct aws_string * hdr) {
+    return (const uint8_t *)(hdr + 1);
+}
+
+static inline const struct aws_string * aws_string_from_literal_new(struct aws_allocator * allocator, const char * literal) {
+    size_t len = strlen(literal);
+    struct aws_string * hdr = aws_mem_acquire(allocator, sizeof(struct aws_string) + len);
+    if (!hdr) {aws_raise_error(AWS_ERROR_OOM); return NULL;}
+    hdr->allocator = allocator;
+    hdr->len = len;
+    memcpy((void *)aws_string_bytes(hdr), literal, len);
+    return hdr;
+}
+
+static inline const struct aws_string * aws_string_from_array_new(struct aws_allocator * allocator, const uint8_t * bytes, size_t len) {
+    struct aws_string * hdr = aws_mem_acquire(allocator, sizeof(struct aws_string) + len);
+    if (!hdr) {aws_raise_error(AWS_ERROR_OOM); return NULL;}
+    hdr->allocator = allocator;
+    hdr->len = len;
+    memcpy((void *)aws_string_bytes(hdr), bytes, len);
+    return hdr;
+}
+
+static inline const struct aws_string * aws_string_from_c_str_new(struct aws_allocator * allocator, const char * bytes, size_t len) {
```

Comment:
This is identical to aws_string_from_array_new only with a `const char *`. Consider using `const void *` instead; or at least delegate from one to the other.

Reply:
This was not my design decision. The different constructors for aws_string were the same as what we already had for aws_byte_buf. I have noticed that we have warnings on strongly enough that trying to call the function that expects unsigned characters on arrays of signed characters, or the reverse, causes an error, which is probably why this was built this way in the first place. I don't have a super strong feeling about what the right set of constructor functions is, but it does seem to me that they should be the same for both aws_byte_buf and aws_string. Opinions @JonathanHenson ?

Reply:
On further thought, I think this is a better naming convention:

aws_string_from_literal_new -> aws_string_from_c_str_new
aws_string_from_array_new -> aws_string_from_uint8_arr_new
aws_string_from_c_str_new-> aws_string_from_int8_arr_new

To me "c_str" means null terminated string, which is how we have been using the term literal before.

Reply:
Also, I will do a test of whether making the bytes argument in the last two functions a void * obviates the need for two separate functions for the signed and unsigned cases.

Reply:
Yes, it looks like @bdonlan was right, and we can avoid making separate functions for unsigned and signed character arrays by making the arguments (void *).

New constructors where we don't specify length but get it from strlen will be aws_string_from_c_str_new and aws_byte_buf_from_c_str. (What used to be _literal.) New constructors where we do specify length will be aws_string_from_array_new and aws_byte_buf_from_array. These changes will be included on the next PR update.